src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -750,5 +733,304 @@ private static string ToHtmlCase(string name)
         {
             return HtmlCaseRegex.Replace(name, HtmlCaseRegexReplacement).ToLowerInvariant();
         }
+
+        // Internal for testing
+        internal class RequiredAttributeParser
+        {
+            private static readonly IReadOnlyDictionary<char, TagHelperRequiredAttributeValueComparison> CssValueComparisons =
+                new Dictionary<char, TagHelperRequiredAttributeValueComparison>
+                {
+                    { '=', TagHelperRequiredAttributeValueComparison.FullMatch },
+                    { '^', TagHelperRequiredAttributeValueComparison.PrefixMatch },
+                    { '$', TagHelperRequiredAttributeValueComparison.SuffixMatch }
+                };
+            private static readonly char[] InvalidPlainAttributeNameCharacters = { ' ', '\t', ',', RequiredAttributeWildcardSuffix };
+            private static readonly char[] InvalidCssAttributeNameCharacters = (new[] { ' ', '\t', ',', ']' })
+                .Concat(CssValueComparisons.Keys)
+                .ToArray();
+            private static readonly char[] InvalidCssQuotelessValueCharacters = { ' ', '\t', ']' };
+
+            private int _index;
+            private string _requiredAttributes;
+
+            public RequiredAttributeParser(string requiredAttributes)
+            {
+                _requiredAttributes = requiredAttributes;
+            }
+
+            private char Current => _requiredAttributes[_index];
+
+            private bool AtEnd => _index >= _requiredAttributes.Length;
+
+            public bool TryParse(
+                ErrorSink errorSink,
+                out IEnumerable<TagHelperRequiredAttributeDescriptor> requiredAttributeDescriptors)
+            {
+                if (string.IsNullOrEmpty(_requiredAttributes))
+                {
+                    requiredAttributeDescriptors = Enumerable.Empty<TagHelperRequiredAttributeDescriptor>();
+                    return true;
+                }
+
+                requiredAttributeDescriptors = null;
+                var descriptors = new List<TagHelperRequiredAttributeDescriptor>();
+
+                while (!AtEnd)
+                {
+                    PassOptionalWhitespace();
+
+                    TagHelperRequiredAttributeDescriptor descriptor;
+                    if (At('['))
+                    {
+                        descriptor = ParseCssSelector(errorSink);
+                    }
+                    else
+                    {
+                        descriptor = ParsePlainSelector(errorSink);
```

Comment:
Weird to get an error about name length here. Would be silent it `AtEnd` before moving past whitespace. Why is trailing whitespace an error at all?